README
caer
====

AER event-based framework, written in C, targeting embedded systems.

REQUIREMENTS:

Linux, MacOS X or Windows (for Windows build instructions see README.Windows)
cmake >= 2.6
gcc >= 5.2 or clang >= 3.6
libcaer >= 2.2.0
mini-xml (mxml) >= 2.7
libuv >= 1.7.5
Boost >= 1.50 (with system, filesystem, program_options)
Optional: libpng >= 1.6 (input/output frame PNG compression)
Optional: tcmalloc >= 2.2 (faster memory allocation)
Optional: allegro5 >= 5.0.11 (visualizer module)
Optional: OpenCV >= 3.1 (cameracalibration, poseestimation modules)

INSTALLATION:

1) configure: 

$ cmake <OPTIONS> <MODULES_TO_BUILD> .

The following options are currently supported:
-DUSE_TCMALLOC=1 -- Enables usage of TCMalloc from Google to allocate memory.

The following modules can currently be selected to be built:
-DDVS128=1 -- DVS128 device input.
-DEDVS=1 -- eDVS4337 device input.
-DDAVIS=1 -- DAVIS device input.
-DDYNAPSE=1 -- Dynap-se device input (neuromorphic chip).
-DBAFILTER=1 -- Filter background activity (uncorrelated noise).
-DFRAMEENHANCER=1 -- Demosaic/enhance frames.
-DCAMERACALIBRATION=1 -- Calculate and apply single camera lens calibration.
-DPOSEESTIMATION=1 -- Estimate pose of camera relative to special markers.
-DSTATISTICS=1 -- Print statistics to console.
-DVISUALIZER=1 -- Open windows in which to visualize data.
-DINPUT_FILE=1 -- Get input from an AEDAT file.
-DOUTPUT_FILE=1 -- Write data to an AEDAT 3.X file.
-DINPUT_NETWORK=1 -- Read input from a network stream.
-DOUTPUT_NETWORK=1 -- Send data out via network.
-DROTATE=1 -- Rotate events.
-DMEDIANTRACKER=1 -- Track points of high event activity.
-DRECTANGULARTRACKER=1 -- Track clusters of events.
-DDYNAMICRECTANGULARTRACKER=1 -- Track a variable number of clusters of events.
-DSPIKEFEATURES=1 -- Create frames which represents decay of polarity events.
-DMEANRATEFILTER=1 -- Measure mean rate of spike events.
-DSYNAPSERECONFIG=1 -- Enable Davis240C to Dynap-se mapping 
-DFPGASPIKEGEN=1 -- Enable FPGA spike generator Dynap-se
-DPOISSONSPIKEGEN=1 -- Enable FPGA Poisson spike generator for Dynap-se

To enable all just type:
 cmake -DDVS128=1 -DEDVS=1 -DDAVIS=1 -DDYNAPSE=1 -DBAFILTER=1 -DFRAMEENHANCER=1 -DCAMERACALIBRATION=1  
 -DPOSEESTIMATION=1 -DSTATISTICS=1  -DVISUALIZER=1 -DINPUT_FILE=1 -DOUTPUT_FILE=1 -DINPUT_NETWORK=1  
 -DOUTPUT_NETWORK=1 -DROTATE=1 -DMEDIANTRACKER=1  -DRECTANGULARTRACKER=1 -DDYNAMICRECTANGULARTRACKER=1  
 -DSPIKEFEATURES=1  -DMEANRATEFILTER=1 -DSYNAPSERECONFIG=1 -DFPGASPIKEGEN=1 -DPOISSONSPIKEGEN=1 .

2) build:

$ make

3) install:

$ make install

USAGE:

You will need a 'caer-config.xml' file that specifies which and how modules
should be interconnected. A good starting point is 'docs/davis-config.xml', 
or 'docs/dynapse-config.xml', please also read through 'docs/modules.txt' for 
an explanation of the modules system and its configuration syntax.

$ caer-bin (see docs/ for more info on how to use cAER)
$ caer-ctl (command-line run-time control program, optional)




ChangeLog
Release 1.2.0 - 23.03.2018
INCOMPATIBLE CHANGES
- The modules are now fully separated from the core system by a well-defined
  API in /usr/include/caer-sdk/ and a library called libcaersdk. Together
  with the shared CMake module 'caer-base.cmake' installed by libcaer 2.4.2,
  it is now possible to easily build modules outside of the cAER source tree.
  All Dynap-SE related modules have been moved to their own external repo at
  https://gitlab.com/aiCTX/caer-modules-dynapse/ and can now be managed
  independently. Only core system and vision modules remain in the main repo.
  This change will require updates in your modules source code to use the
  new header locations, in most cases:

 #include "main.h"
 #include "base/mainloop.h"
 #include "base/module.h"

  should be substituted with only:

 #include <caer-sdk/mainloop.h>

  Following a full list of the headers that changed location:

 "base/mainloop.h"         ->  <caer-sdk/mainloop.h>
 "base/module.h"           ->  <caer-sdk/module.h>
 "main.h"                  ->  <caer-sdk/utils.h>
 "ext/buffers.h"           ->  <caer-sdk/buffers.h>
 "ext/portable_time.h"     ->  <caer-sdk/cross/portable_time.h>
 "ext/portable_misc.h"     ->  <caer-sdk/cross/portable_misc.h>
 "ext/pathmax.h"           ->  <caer-sdk/cross/pathmax.h>
 "ext/c11threads_posix.h"  ->  <caer-sdk/cross/c11threads_posix.h>
 "ext/threads_ext.h"       ->  <caer-sdk/cross/threads_ext.h>
 "ext/sshs/sshs.h"         ->  <caer-sdk/sshs/sshs.h>
 "ext/sshs/sshs.hpp"       ->  <caer-sdk/sshs/sshs.hpp>

  Anything else you might have been using that resides in the ext/ directory
  is fundamentally forbidden for external modules. The remaining headers
  there will soon disappear, you will have to copy them into your module and
  keep a private copy, or search for suitable alternatives.
- All modules that can be enabled are now enabled by default, so most -Dxxx
  compilation control switches fall away. Only -DUSE_OPENCV=1/0 and
  -DVISUALIZER=1/0 remain to control the inclusion of external dependencies
  (OpenCV for cameracalibration, framestatistics - SFML for visualizer).

NEW FEATURES
- Docs: added two camera example 'davis-2cams-config.xml'.

BUG FIXES
- Output modules: properly exit on initialization failure.
- Visualizer: removed 'None' event handler/renderer option from being
  selectable in the GUI.
- Mainloop/Modules: do basic checks on modules at module discovery time,
  so that broken modules don't even appear in the list of available ones.


Release 1.1.0 - 05.12.2017
INCOMPATIBLE CHANGES
- Dynap-SE: biasing has been cleaned up, please use the functions
  caerDynapseGetBiasCore() and caerDynapseSetBiasCore() from dynapse_utils.h
  to get/set biases from other modules. No other functions to do this
  are supported. The configuration path for biases changed to be more
  clearly laid out and is now: 'bias/Ux/Cy/<BiasName>/'.
  Also the software spike generator is now its own module, called
  'softspikegen'. The 'fpgaspikegen' and 'poissonspikegen' modules have
  been ported to the new modules system. The 'monitorneufilter' module is
  now directly integrated as part of the main Dynap-SE device input module.
- Modules: we are shifting to a SDK-like system. Most non-core modules
  have been moved to a private repository. Please contact us for access.
- Dependencies: requires libcaer 2.4.0. mini-xml is not needed anymore,
  and libuv is only needed for the misc/out/ modules. Removal of this
  dependency is tracked in issue #73. Also the Boost.iostreams package
  is now required, in addition to system, filesystem and program_options.
  The SLRE dependency in SSHS has also been removed in favor of std::regex.
- License: changed license from GPL-2 to LGPL-2.1 to allow for closed
  source plugins/modules.
- Devices/Inputs: removed 'highestTimestamp' sourceInfo node attribute for
  input devices and modules. It was not being used anywhere.

NEW FEATURES
- Documentation: added several XML configuration examples in docs/.
- Modules: added 'description' field to 'struct caer_module_info'.
- Modules: modules can now wait, using caerMainloopGetModuleInputIDs(),
  for their inputs to be ready.
- SSHS: rewritten in C++, added sshsAttributeReadModifier callback
  to influence values returned when reading attributes.
- SSHS: added sshsNodeCreateAttributePollTime(), sshsNodeCreateAttributeListOptions()
  and sshsNodeCreateAttributeFileChooser() functions to create
  informative attributes, to be used by GUIs and clients.
- Visualizer: rewritten to use C++ and the SFML libraries. This now
  works on MacOS X too. Fixes issue #92.
- mainloop.cpp: order modules in such a way to minimize event packets
  needing to be copied around (performance improvement).
- mainloop.h: added caerMainloopGetModuleNode() function.
- netparser: network parser module for Dynap-SE.

BUG FIXES
- mainloop.cpp: fixed various issues with module order generation.
  Fixes issues #121, #112, #107.
- sshs_node.cpp: fixed various memory problems, issues #120, #119,
  #118, #117, and #93.
- Boost: improved compatibility with older 1.5X versions.
- module.cpp: changed separation character for multiple modules search
  paths in the 'modulesSearchPath' attribute to '|' instead of ':'.
  This is needed for Windows compatibility.
- Input modules: correctly recognize Dynap-SE files, fixes issue #128.


Release 1.0.1 - 03.07.2017 (ChangeLog incomplete)
INCOMPATIBLE CHANGES
- Visualizer: cannot work on MacOS X due to how the Allegro5 library
  implementes initialization by overriding main(). Trying to enable
  the visualizer on MacOS X results in an error now.

NEW FEATURES
- Docs: new 'docs/modules.txt' document to describe modules system.
- Input/Output modules: use caerModuleLog() where possible.

main.c
/*
 * main.c
 *
 *  Created on: Oct 6, 2013
 *      Author: chtekk
 *
 *  Compile & run:
 *  $ cd caer/
 *  $ rm -rf CMakeFiles CMakeCache.txt
 *  $ CC=clang-3.7 cmake [-DJAER_COMPAT_FORMAT=1 -DENABLE_VISUALIZER=1 -DENABLE_NET_STREAM=1] -DDAVISFX2 .
 *  $ make
 *  $ ./caer-bin
 */

#include "main.h"
#include "base/config.h"
#include "base/config_server.h"
#include "base/log.h"
#include "base/mainloop.h"
#include "base/misc.h"

// Devices support.
#ifdef DVS128
#include "modules/ini/dvs128.h"
#endif
#ifdef DAVISFX2
#include "modules/ini/davis_fx2.h"
#endif
#ifdef DAVISFX3
#include "modules/ini/davis_fx3.h"
#endif

// Input/Output support.
#ifdef ENABLE_FILE_INPUT
#include "modules/misc/in/file.h"
#endif
#ifdef ENABLE_NETWORK_INPUT
#include "modules/misc/in/net_tcp.h"
#include "modules/misc/in/unix_socket.h"
#endif

#ifdef ENABLE_FILE_OUTPUT
#include "modules/misc/out/file.h"
#endif
#ifdef ENABLE_NETWORK_OUTPUT
#include "modules/misc/out/net_tcp_server.h"
#include "modules/misc/out/net_tcp.h"
#include "modules/misc/out/net_udp.h"
#include "modules/misc/out/unix_socket_server.h"
#include "modules/misc/out/unix_socket.h"
#endif

// Common filters support.
#ifdef ENABLE_BAFILTER
#include "modules/backgroundactivityfilter/backgroundactivityfilter.h"
#endif
#ifdef ENABLE_CAMERACALIBRATION
#include "modules/cameracalibration/cameracalibration.h"
#endif
#ifdef ENABLE_FRAMEENHANCER
#include "modules/frameenhancer/frameenhancer.h"
#endif
#ifdef ENABLE_STATISTICS
#include "modules/statistics/statistics.h"
#endif
#ifdef ENABLE_VISUALIZER
#include "modules/visualizer/visualizer.h"
#endif
#ifdef ENABLE_INFOFILTER
#include "modules/infofilter/infofilter.h"
#endif
#ifdef ENABLE_RECTANGULARTRACKER
#include "modules/rectangulartracker/rectangulartracker.h"
#endif
#ifdef ENABLE_RECTANGULARTRACKER_DYNAMIC
#include "modules/rectangulartracker_dynamic/rectangulartracker_dynamic.h"
#endif
#ifdef ENABLE_RECTANGULARTRACKER_PI
#include "modules/rectangulartracker_pi/rectangulartracker_pi.h"
#endif
#ifdef ENABLE_MEDIANTRACKER
#include "modules/mediantracker/mediantracker.h"
#endif
#ifdef ENABLE_MEANRATEFILTER_DVS
#include <libcaer/events/frame.h>
#include "modules/meanratefilter_dvs/meanratefilter_dvs.h"
#endif
#ifdef ENABLE_ROTATEFILTER
#include "modules/rotatefilter/rotatefilter.h"
#endif
#ifdef ENABLE_ACTIVITYINDICATOR
#include "modules/activityindicator/activityindicator.h"
#endif
#ifdef ENABLE_OPENCVDISPLAY
#include "modules/opencvdisplay/opencvdisplay_module.h"
#endif
#ifdef ENABLE_OPENCVOPTICFLOW
#include "modules/opencvopticflow/opticflow.h"
#endif
#ifdef ENABLE_DENOISINGAUTOENCODER
#include "modules/denoisingautoencoder/denoisingautoencoder_module.h"
#endif
#ifdef ENABLE_PIXELMATRIX
#include "modules/pixelmatrix/pixelmatrix.h"
#endif
#ifdef ENABLE_DEPRESSINGSYNAPSEFILTER
#include "modules/depressingsynapsefilter/depressingsynapsefilter.h"
#endif
#ifdef ENABLE_SPIKEFEATURES
#include "modules/spikefeatures/spikefeatures.h"
#endif
#ifdef ENABLE_SPIRALVIEW
#include "modules/spiralview/spiralview.h"
#endif

#ifdef ENABLE_IMAGEGENERATOR
#include "modules/imagegenerator/imagegenerator.h"
#define CLASSIFYSIZE 180
#define DISPLAYIMGSIZE 180
#endif
#ifdef ENABLE_CAFFEINTERFACE
#define CAFFEVISUALIZERSIZE 1024 
#include "modules/caffeinterface/wrapper.h"
#endif
#ifdef ENABLE_NULLHOPINTERFACE
#include "modules/nullhopinterface/nullhopinterface.h"
#endif

static bool mainloop_1(void);

static bool mainloop_1(void) {
	// An eventPacketContainer bundles event packets of different types together,
	// to maintain time-coherence between the different events.
	caerEventPacketContainer container = NULL;
	caerSpecialEventPacket special = NULL;
	caerPolarityEventPacket polarity = NULL;

#if (defined(ENABLE_FILE_INPUT) || defined(ENABLE_NETWORK_INPUT)) && (defined(DAVISFX2) || defined(DAVISFX3))
#error "Only one input mode can be enabled: please choose DAVISFX2/3 or ENABLE_FILE_INPUT"
#endif

#ifdef ENABLE_VISUALIZER
	caerVisualizerEventHandler visualizerEventHandler = NULL;
#endif

	// Input modules grab data from outside sources (like devices, files, ...)
	// and put events into an event packet.
#ifdef DVS128
	container = caerInputDVS128(1);

	// Typed EventPackets contain events of a certain type.
	special = (caerSpecialEventPacket) caerEventPacketContainerGetEventPacket(container, SPECIAL_EVENT);
	polarity = (caerPolarityEventPacket) caerEventPacketContainerGetEventPacket(container, POLARITY_EVENT);
#endif

#ifdef DAVISFX2
	container = caerInputDAVISFX2(1);
#endif
#ifdef DAVISFX3
	container = caerInputDAVISFX3(1);
#endif
#if defined(DAVISFX2) || defined(DAVISFX3)
	// Typed EventPackets contain events of a certain type.
	special = (caerSpecialEventPacket) caerEventPacketContainerGetEventPacket(container, SPECIAL_EVENT);
	polarity = (caerPolarityEventPacket) caerEventPacketContainerGetEventPacket(container, POLARITY_EVENT);

	// Frame and IMU events exist only with DAVIS cameras.
	caerFrameEventPacket frame = NULL;
	caerIMU6EventPacket imu = NULL;

	frame = (caerFrameEventPacket) caerEventPacketContainerGetEventPacket(container, FRAME_EVENT);
	imu = (caerIMU6EventPacket) caerEventPacketContainerGetEventPacket(container, IMU6_EVENT);
#endif

#ifdef ENABLE_FILE_INPUT
	container = caerInputFile(10);
#endif
#ifdef ENABLE_NETWORK_INPUT
	container = caerInputNetTCP(11);
#endif
#if defined(ENABLE_FILE_INPUT) || defined(ENABLE_NETWORK_INPUT)
#ifdef ENABLE_VISUALIZER
	visualizerEventHandler = &caerInputVisualizerEventHandler;
#endif

	// Typed EventPackets contain events of a certain type.
	// We search for them by type here, because input modules may not have all or any of them.
	special = (caerSpecialEventPacket) caerEventPacketContainerFindEventPacketByType(container, SPECIAL_EVENT);
	polarity = (caerPolarityEventPacket) caerEventPacketContainerFindEventPacketByType(container, POLARITY_EVENT);

	caerFrameEventPacket frame = NULL;
	caerIMU6EventPacket imu = NULL;

	frame = (caerFrameEventPacket) caerEventPacketContainerFindEventPacketByType(container, FRAME_EVENT);
	imu = (caerIMU6EventPacket) caerEventPacketContainerFindEventPacketByType(container, IMU6_EVENT);
#endif

	// Filter that adds buttons and timer for recording data
	// or playing a recording file. It implements fast forward and
	// slow motion buttons, as well as play again from start.
	// It is a very basic camera interface (320x240)
#ifdef ENABLE_INFOFILTER
#if defined(ENABLE_FILE_INPUT) && defined(ENABLE_FILE_OUTPUT)
	caerInfoFilter(78, container, 10, 7);
#elif defined(ENABLE_FILE_INPUT) && !defined(ENABLE_FILE_OUTPUT)
	caerInfoFilter(78, container, 10, NULL);
#elif !defined(ENABLE_FILE_INPUT) && defined(ENABLE_FILE_OUTPUT)
	caerInfoFilter(78, container, NULL, 7);
#endif
#endif

	// Filters process event packets: for example to suppress certain events,
	// like with the Background Activity Filter, which suppresses events that
	// look to be uncorrelated with real scene changes (noise reduction).
#ifdef ENABLE_BAFILTER
	caerBackgroundActivityFilter(2, polarity);
#endif

	// Filters can also extract information from event packets: for example
	// to show statistics about the current event-rate.
#ifdef ENABLE_STATISTICS
	caerStatistics(3, (caerEventPacketHeader) polarity, 1000);
#endif

	// Filter that rotate events in different ways
#ifdef ENABLE_ROTATEFILTER
	caerRotateFilter(16, polarity);
#endif

	// Filters out rapidly firing input using depressing probabilistic synapse model
#ifdef ENABLE_DEPRESSINGSYNAPSEFILTER
	caerDepressingSynapseFilter(25, polarity);
#endif

	// Filters that track multiple objects by using rectangular clusters (fixed length array version)
#ifdef ENABLE_RECTANGULARTRACKER
	caerFrameEventPacket rectangularFrame = NULL;
	caerRectangulartrackerFilter(12, polarity, &rectangularFrame);
#endif

	// Filters that track multiple objects by using rectangular clusters (dynamic linked list version)
#ifdef ENABLE_RECTANGULARTRACKER_DYNAMIC
	caerFrameEventPacket rectangularDynamicFrame = NULL;
	caerRectangulartrackerDynamicFilter(120, polarity, &rectangularDynamicFrame);
#endif

	// Rectangular tracker for raspberry pi
#ifdef ENABLE_RECTANGULARTRACKER_PI
	caerRectangulartrackerPiFilter(121, polarity);
#endif

	// Filter that track one object by using the median position information
#ifdef ENABLE_MEDIANTRACKER
	caerFrameEventPacket medianFrame = NULL;
	caerMediantrackerFilter(13, polarity, &medianFrame);
#endif

	// Show how crowed the areas is (Mensa project)
#ifdef ENABLE_ACTIVITYINDICATOR
	caerFrameEventPacket activityFrame = NULL;
	AResults rr = caerActivityIndicator(17, polarity, &activityFrame);
#endif

#if defined(ENABLE_OPENCVDISPLAY) && defined(ENABLE_ACTIVITYINDICATOR)
	caerFrameEventPacket frameRes = NULL;
	if(rr->activityValue != -1){
		frameRes = caerOpenCVDisplay(18, rr);
	}
#endif

	// Filter that show the mean rate of events
#ifdef ENABLE_MEANRATEFILTER_DVS
	caerFrameEventPacket freqplot = NULL;
	caerMeanRateFilterDVS(15, polarity, &freqplot);
#endif

	// Enable APS frame image enhancements.
#ifdef ENABLE_FRAMEENHANCER
	frame = caerFrameEnhancer(4, frame);
#endif

	// Enable image and event undistortion by using OpenCV camera calibration.
#ifdef ENABLE_CAMERACALIBRATION
	caerCameraCalibration(5, polarity, frame);
#endif

	//Enable camera pose estimation
#ifdef ENABLE_POSEESTIMATION
	caerPoseCalibration(6, polarity, frame);
#endif

#ifdef ENABLE_SPIKEFEATURES
	caerFrameEventPacket spikeFeatures = NULL;
	caerSpikeFeatures(26, polarity, &spikeFeatures);
#endif


	// A simple visualizer exists to show what the output looks like.
#ifdef ENABLE_VISUALIZER
	caerVisualizer(60, "Polarity", &caerVisualizerRendererPolarityEvents, visualizerEventHandler, (caerEventPacketHeader) polarity);
#if defined(ENABLE_FILE_INPUT)
	if (frame != NULL) {
		caerVisualizer(61, "Frame", &caerVisualizerRendererFrameEvents, visualizerEventHandler, (caerEventPacketHeader) frame);
	}
	if (imu != NULL) {
		caerVisualizer(62, "IMU6", &caerVisualizerRendererIMU6Events, visualizerEventHandler, (caerEventPacketHeader) imu);
	}
#endif
#if defined(DAVISFX2) || defined(DAVISFX3)
	caerVisualizer(61, "Frame", &caerVisualizerRendererFrameEvents, visualizerEventHandler, (caerEventPacketHeader) frame);
	caerVisualizer(62, "IMU6", &caerVisualizerRendererIMU6Events, visualizerEventHandler, (caerEventPacketHeader) imu);
#endif
#ifdef ENABLE_MEANRATEFILTER_DVS
	if(freqplot != NULL){
		caerVisualizer(70, "MeanRateFrequency", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) freqplot);
	}
#endif
#if defined(ENABLE_OPENCVDISPLAY) && defined(ENABLE_ACTIVITYINDICATOR)
	caerVisualizer(71, "FrameRes", &caerVisualizerRendererFrameEvents, visualizerEventHandler, (caerEventPacketHeader) frameRes);
#endif
	//caerVisualizerMulti(68, "PolarityAndFrame", &caerVisualizerMultiRendererPolarityAndFrameEvents, visualizerEventHandler, container);
#ifdef	ENABLE_SPIKEFEATURES
	if(spikeFeatures != NULL){
		caerVisualizer(74, "SpikeSurface", &caerVisualizerRendererFrameEvents, visualizerEventHandler, (caerEventPacketHeader) spikeFeatures);
	}
#endif
#endif

#ifdef ENABLE_FILE_OUTPUT
	// Enable output to file (AEDAT 3.X format).
#ifdef DVS128
	caerOutputFile(7, 2, polarity, special);
#else
	caerOutputFile(7, 4, polarity, frame, imu, special);
#endif
#endif

#ifdef ENABLE_NETWORK_OUTPUT
	// Send polarity packets out via TCP. This is the server mode!
	// External clients connect to cAER, and we send them the data.
	// WARNING: slow clients can dramatically slow this and the whole
	// processing pipeline down!
#ifdef DVS128
	caerOutputNetTCPServer(8, 2, polarity, special);
#else
	caerOutputNetTCPServer(8, 4, polarity, frame, imu, special);
#endif

	// And also send them via UDP. This is fast, as it doesn't care what is on the other side.
#ifdef DVS128
	caerOutputNetUDP(9, 2, polarity, special);
#else
	caerOutputNetUDP(9, 4, polarity, frame, imu, special);
#endif
#endif

#ifdef ENABLE_IMAGEGENERATOR
	// it creates images by accumulating spikes
	int * classifyhist = calloc((int)CLASSIFYSIZE * CLASSIFYSIZE, sizeof(int));
	if (classifyhist == NULL) {
			return (false); // Failure.
	}
	bool *haveimage;
	haveimage = (bool*)malloc(1);
	unsigned char ** frame_img_ptr = calloc(sizeof(unsigned char *), 1);
	// generate image and place it in classifyhist
	caerImageGenerator(20, polarity, CLASSIFYSIZE, classifyhist, haveimage);
#ifdef ENABLE_VISUALIZER
   //put image into a frame packet containing a single frame
   caerFrameEventPacket imagegeneratorFrame = NULL;
   if(haveimage[0]){
	  caerImageGeneratorMakeFrame(20, classifyhist, &imagegeneratorFrame, CLASSIFYSIZE);
   }
#endif
#endif

#if defined(ENABLE_SPIRALVIEW) && defined(ENABLE_IMAGEGENERATOR)
	// it creates images by accumulating spikes
	int * spiralview = calloc((int)CLASSIFYSIZE * CLASSIFYSIZE * 3, sizeof(int));
	if (spiralview == NULL) {
			return (false); // Failure.
	}

	caerFrameEventPacket spiralFrame = NULL;
	if(haveimage[0]){
		caerSpiralView(27, polarity, CLASSIFYSIZE, classifyhist, spiralview, &spiralFrame);
	}
#endif


	// this modules requires image generator
#if defined(ENABLE_CAFFEINTERFACE) || defined(ENABLE_NULLHOPINTERFACE)
	// this wrapper let you interact with caffe framework
	// we now classify the latest image that has been generated by the imagegenerator filter
	// it also can create an image with activations values for different layers
        caerFrameEventPacket networkActivity = NULL; /* visualization of network activity */
        char * classification_results = (char *)calloc(1024, sizeof(char));/* classification_results: */
        if (classification_results == NULL) {
                caerLog(CAER_LOG_ERROR, __func__, "Failed to allocate classification_results.");
                return (false);
        }else{
                caerMainloopFreeAfterLoop(&free, classification_results);
        }
        int * classification_results_id;
        classification_results_id = (int*)malloc(1);    /* classification_results: */
        classification_results_id[0] = -1; // init
        if (classification_results_id == NULL) {
                caerLog(CAER_LOG_ERROR, __func__, "Failed to allocate classification_results_id.");
                return (false);
        }else{
                caerMainloopFreeAfterLoop(&free, classification_results_id);
        }
	
#ifdef ENABLE_IMAGEGENERATOR
        if(haveimage[0]) {
                caerCaffeWrapper(21, classifyhist, CLASSIFYSIZE, classification_results, classification_results_id,
                        &networkActivity, CAFFEVISUALIZERSIZE);
        }
#endif
#endif

        // add classification results to the image generator frame
#ifdef ENABLE_IMAGEGENERATOR
#if defined(ENABLE_CAFFEINTERFACE)
        if(haveimage[0]){
                caerImageGeneratorAddText(23, classifyhist, &imagegeneratorFrame, CLASSIFYSIZE);
        }
#endif
#endif

#ifdef ENABLE_PIXELMATRIX
#if defined(ENABLE_CAFFEINTERFACE)
       if(haveimage[0]){
    	   caerPixelMatrixFilter(24, classification_results_id);
       }
#endif
#endif

#ifdef ENABLE_NULLHOPINTERFACE
	caerNullHopWrapper(22, classifyhist, haveimage, classification_results);
#endif

#if defined(ENABLE_OPENCVOPTICFLOW) && defined(ENABLE_IMAGEGENERATOR) && defined(ENABLE_VISUALIZER)
	caerFrameEventPacket frameFlow = NULL;
	//caerFrameEventPacket frameFlow = caerOpticFlow(19, imagegeneratorFrame);
#endif

#if defined(ENABLE_DENOISINGAUTOENCODER) && defined(ENABLE_IMAGEGENERATOR) && defined(ENABLE_VISUALIZER)
	caerFrameEventPacket frameAutoEncoderFeatures = NULL;
	frameAutoEncoderFeatures = caerDenAutoEncoder(24, imagegeneratorFrame);
#endif

#if defined(ENABLE_VISUALIZER) && defined (ENABLE_IMAGEGENERATOR) && defined(ENABLE_SPIRALVIEW)
	 if(haveimage[0]){
		 caerVisualizer(88, "SpiralView", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) spiralFrame);
	}
#endif

#if defined(ENABLE_VISUALIZER) && defined (ENABLE_IMAGEGENERATOR)
	caerVisualizer(65, "ImageStreamerHist", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) imagegeneratorFrame);
#ifdef ENABLE_CAFFEINTERFACE
	//show the activations of the deep network
	caerVisualizer(66, "DeepNetworkActivations", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) networkActivity);
#endif
#endif

#if defined(ENABLE_RECTANGULARTRACKER) && defined (ENABLE_VISUALIZER)
	caerVisualizer(67, "ImageClusters", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) rectangularFrame);
#endif

#if defined(ENABLE_RECTANGULARTRACKER_DYNAMIC) && defined (ENABLE_VISUALIZER)
	caerVisualizer(670, "ImageClusters", &caerVisualizerRendererFrameEvents, &caerVisualizerEventHandlerFrameEvents, (caerEventPacketHeader) rectangularDynamicFrame);
#endif

#if defined(ENABLE_MEDIANTRACKER) && defined (ENABLE_VISUALIZER)
	caerVisualizer(68, "ImageMedian", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) medianFrame);
#endif

#if defined(ENABLE_ACTIVITYINDICATOR) && defined (ENABLE_VISUALIZER)
	caerVisualizer(69, "ImageActivity", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) activityFrame);
#endif

#if defined(ENABLE_OPENCVOPTICFLOW) && defined(ENABLE_IMAGEGENERATOR)
	caerVisualizer(72, "OpticFlow", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) frameFlow);
#endif

#if defined(ENABLE_DENOISINGAUTOENCODER) && defined(ENABLE_IMAGEGENERATOR) && defined(ENABLE_VISUALIZER)
	caerVisualizer(73, "DenoiserAutoEncoder", &caerVisualizerRendererFrameEvents, NULL, (caerEventPacketHeader) frameAutoEncoderFeatures);
#endif

	return (true); // If false is returned, processing of this loop stops.
}

int main(int argc, char **argv) {

	// Set thread name.
	thrd_set_name("Main");

	// Initialize config storage from file, support command-line overrides.
	// If no init from file needed, pass NULL.
	caerConfigInit("caer-config.xml", argc, argv);

	// Initialize logging sub-system.
	caerLogInit();

	// Daemonize the application (run in background, NOT AVAILABLE ON WINDOWS).
	// caerDaemonize();

	// Initialize visualizer framework (load fonts etc.).
#ifdef ENABLE_VISUALIZER
	caerVisualizerSystemInit();
#endif

	// Start the configuration server thread for run-time config changes.
	caerConfigServerStart();

	// Finally run the main event processing loops.
	struct caer_mainloop_definition mainLoops[1] = { { 1, &mainloop_1 } };
	caerMainloopRun(&mainLoops, 1); // Only start Mainloop 1.

	// After shutting down the mainloops, also shutdown the config server
	// thread if needed.
	caerConfigServerStop();

	return (EXIT_SUCCESS);
}
